source/math/FloatMatrix.ooc

```diff
@@ -282,59 +330,76 @@ FloatMatrix : cover {
+		for (x in 0 .. otherWidth) {
+			for (y in 0 .. t height) {
+				temp := result elements[x + y * otherWidth]
+				for (z in 0 .. t width)
+					temp += this elements[z + y * t width] * other elements[x + z * otherWidth]
+				result elements[x + y * otherWidth] = temp
 			}
 		}
+		if (this elements != other elements)
+			other free(Owner Receiver)
 		this free(Owner Receiver)
-		other free(Owner Receiver)
 		result
 	}
 	operator + (other: This) -> This {
-		if (this dimensions != other dimensions)
-			raise("Invalid dimensions in FloatMatrix + operator: dimensions must match!")
-		result := This new(this dimensions)
-		for (i in 0 .. this dimensions area)
+		t := this take()
+		version(safe) {
+			if (t dimensions != other take() dimensions)
+				raise("Invalid dimensions in FloatMatrix + operator: dimensions must match!")
+		}
+		result := This new(t dimensions)
+		for (i in 0 .. t dimensions area)
 			result elements[i] = this elements[i] + other elements[i]
+		if (this elements != other elements)
+			other free(Owner Receiver)
 		this free(Owner Receiver)
-		other free(Owner Receiver)
 		result
 	}
 	operator - (other: This) -> This {
-		if (this dimensions != other dimensions)
-			raise("Invalid dimensions in FloatMatrix - operator: dimensions must match!")
-		result := This new(this dimensions)
-		for (i in 0 .. this dimensions area)
+		t := this take()
+		version(safe) {
+			if (t dimensions != other take() dimensions)
+				raise("Invalid dimensions in FloatMatrix - operator: dimensions must match!")
+		}
+		result := This new(t dimensions)
+		for (i in 0 .. t dimensions area)
 			result elements[i] = this elements[i] - other elements[i]
+		if (this elements != other elements)
+			other free(Owner Receiver)
 		this free(Owner Receiver)
-		other free(Owner Receiver)
 		result
 	}
 	operator += (other: This) {
-		if (this dimensions != other dimensions)
-			raise("Invalid dimensions in FloatMatrix += operator: dimensions must match!")
-		for (i in 0 .. this dimensions area)
+		version(safe) {
+			if (this dimensions != other dimensions)
+				raise("Invalid dimensions in FloatMatrix += operator: dimensions must match!")
+		}
+		for (i in 0 .. this take() dimensions area)
 			this elements[i] += other elements[i]
 		other free(Owner Receiver)
 	}
 	operator -= (other: This) {
-		if (this dimensions != other dimensions)
-			raise("Invalid dimensions in FloatMatrix -= operator: dimensions must match!")
-		for (i in 0 .. this dimensions area)
+		version(safe) {
+			if (this take() dimensions != other take() dimensions)
+				raise("Invalid dimensions in FloatMatrix -= operator: dimensions must match!")
+		}
+		for (i in 0 .. this take() dimensions area)
 			this elements[i] -= other elements[i]
 		other free(Owner Receiver)
 	}
 
 	operator * (other: Float) -> This {
-		result := This new(this dimensions)
-		for (i in 0 .. this dimensions area)
+		result := This new(this take() dimensions)
+		for (i in 0 .. this take() dimensions area)
```

Comment:
this is necessary if i don't assign it to a variable. `take()` only affects what it returns.